In [1]:
from promptsmith.dspy_init import get_dspy
from promptsmith.tasks.bad_to_good.bulletize_text_2 import BulletizeText
from promptsmith.judges.judge_bullet_structure import JudgeBulletStructure
from promptsmith.refiners import Refiner

dspy, lm = get_dspy()

/Users/yanivgal/dev/ai21/promptsmith/venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


In [2]:
with open('../../data/text_01.txt', 'r') as f:
    text = f.read()

#### original text

In [3]:
print(text)

The first question in this study stated: What is the strength and direction of the relationship between faculty's perceptions of the importance of course design, communication, time management, and technical competency, and their ability to teach online? To address this question, descriptive statistics (mean and standard deviation) by individual competency items of each of the four competency dimensions: course design, course communication, time management, and technical competency,  were reported in Table 2 on page 23. Most of the items in each competency were rated high for both constructs, Importance, and Ability.
The items manage grades online (M = 4.73) and creating online assignments (M = 4.68)

were rated the highest in the course design competency. In the course communication competency, responding to student questions promptly (M = 4.79) and providing feedback on assignments (M = 4.65) were rated the highest. In time management, scheduling time to design the course prior to de

#### bulletizing

In [4]:
bulletize = dspy.ChainOfThought(BulletizeText)
bulletized = bulletize(input_text=text)

In [5]:
print(bulletized.output_text)

# Faculty Perceptions of Online Teaching Competencies

## Research Question 1: Relationship Between Perceptions and Ability
- **Question:** What is the strength and direction of the relationship between faculty's perceptions of the importance of course design, communication, time management, and technical competency, and their ability to teach online?
- Descriptive statistics (mean and standard deviation) reported in Table 2 on page 23.
- Most items rated high for both constructs: Importance and Ability.
- **Course Design Competency:**
  - Manage grades online (M = 4.73)
  - Creating online assignments (M = 4.68)
- **Course Communication Competency:**
  - Responding to student questions promptly (M = 4.79)
  - Providing feedback on assignments (M = 4.65)
- **Time Management Competency:**
  - Scheduling time to design the course prior to delivery (M = 4.65)
  - Spending weekly hours to grade assignments (M = 4.54)
- **Technical Competency:**
  - Complete basic computer operations (M = 4

#### judging

In [6]:
print("Judging bullet structure...")
judge = dspy.Predict(JudgeBulletStructure)
verdict = judge(text=bulletized.output_text)

print(f"Structure score: {verdict.score:.2f}")
print("\nJudge's reasoning:")
print("----------------")
print(verdict.reasoning)
print("\n" + "-"*80 + "\n")

Judging bullet structure...
Structure score: 0.70

Judge's reasoning:
----------------
- The title is appropriate and follows the guidelines.
- The one-line summary is missing; it should be included immediately after the title.
- Each section begins with a meaningful H2 heading, which is correct.
- The bullet points are well-structured, but the section closing lines are missing; each section should have exactly one italic summary line.
- The separators between major sections are correctly formatted with three dashes.
- There is no extraneous text outside the outline.

--------------------------------------------------------------------------------



#### refine

In [10]:
if verdict.score < 0.9:  # If not perfect, refine
    print("Refining based on feedback...")
    refiner = dspy.Predict(Refiner)
    refined = refiner(
        original_text=text,
        current_output=bulletized.output_text,
        task_type="bulletize",
        feedback_type="structure",
        feedback=verdict.reasoning
    )
    print("Re-judging refined output...")
    new_verdict = judge(text=refined.refined_output)
    print(f"New structure score: {new_verdict.score:.2f} (was {verdict.score:.2f})")
else:
    print("Output already has good structure (score > 0.9)")

Refining based on feedback...
Re-judging refined output...
New structure score: 0.85 (was 0.70)


In [11]:
print(refined.reasoning)

To address the feedback, I added a one-line summary immediately after the title to provide a concise overview of the content. I also included closing summary lines for each section to encapsulate the main points discussed, ensuring that each section has exactly one italicized summary line as requested. This enhances the structure and clarity of the output while maintaining the bulletized format.


In [12]:
print(refined.refined_output)

# Faculty Perceptions of Online Teaching Competencies
*This section explores faculty perceptions regarding the importance of various online teaching competencies and their self-reported abilities.*

---

## Research Question 1: Relationship Between Perceptions and Ability
- **Question:** What is the strength and direction of the relationship between faculty's perceptions of the importance of course design, communication, time management, and technical competency, and their ability to teach online?
- Descriptive statistics (mean and standard deviation) reported in Table 2 on page 23.
- Most items rated high for both constructs: Importance and Ability.
- **Course Design Competency:**
  - Manage grades online (M = 4.73)
  - Creating online assignments (M = 4.68)
- **Course Communication Competency:**
  - Responding to student questions promptly (M = 4.79)
  - Providing feedback on assignments (M = 4.65)
- **Time Management Competency:**
  - Scheduling time to design the course prior to de

#### bulletized text

# Faculty Perceptions of Online Teaching Competencies

## Research Question 1: Relationship Between Perceptions and Ability
- **Question:** What is the strength and direction of the relationship between faculty's perceptions of the importance of course design, communication, time management, and technical competency, and their ability to teach online?
- Descriptive statistics (mean and standard deviation) reported in Table 2 on page 23.
- Most items rated high for both constructs: Importance and Ability.
- **Course Design Competency:**
  - Manage grades online (M = 4.73)
  - Creating online assignments (M = 4.68)
- **Course Communication Competency:**
  - Responding to student questions promptly (M = 4.79)
  - Providing feedback on assignments (M = 4.65)
- **Time Management Competency:**
  - Scheduling time to design the course prior to delivery (M = 4.65)
  - Spending weekly hours to grade assignments (M = 4.54)
- **Technical Competency:**
  - Complete basic computer operations (M = 4.74)
  - Navigating within the course in the learning management system (M = 4.67)
- Findings indicate high importance in course design and communication, suggesting a need for more faculty support.
- **Statistical Significance:**
  - Weak positive relationship between perceived Importance and Ability, r(223) = .249, p < .001.
  
This section highlights the strong perceptions of importance in teaching competencies and their correlation with perceived ability.

---

## Research Question 2: Demographic Factors
- **Question:** What demographic factors are related to faculty's perceptions of the importance of course design, communication, time management, and technical competency, and their ability to teach online?
- Demographic data included: gender, academic rank, online teaching methods, and level taught.
- Descriptive statistics presented in Table 5 on page 32, Chapter IV.
- **ANOVA Results:**
  - Significant differences found in gender and level taught regarding importance of online teaching competencies.
  - No significant differences in gender, academic rank, and online teaching methods.
  - Significant differences in level taught based on perceptions of ability to teach online.
- No significant differences found in other demographic factors.

This section provides insights into how demographic factors influence faculty perceptions of online teaching competencies.

#### refined text

# Faculty Perceptions of Online Teaching Competencies
*This section explores faculty perceptions regarding the importance of various online teaching competencies and their self-reported abilities.*

---

## Research Question 1: Relationship Between Perceptions and Ability
- **Question:** What is the strength and direction of the relationship between faculty's perceptions of the importance of course design, communication, time management, and technical competency, and their ability to teach online?
- Descriptive statistics (mean and standard deviation) reported in Table 2 on page 23.
- Most items rated high for both constructs: Importance and Ability.
- **Course Design Competency:**
  - Manage grades online (M = 4.73)
  - Creating online assignments (M = 4.68)
- **Course Communication Competency:**
  - Responding to student questions promptly (M = 4.79)
  - Providing feedback on assignments (M = 4.65)
- **Time Management Competency:**
  - Scheduling time to design the course prior to delivery (M = 4.65)
  - Spending weekly hours to grade assignments (M = 4.54)
- **Technical Competency:**
  - Complete basic computer operations (M = 4.74)
  - Navigating within the course in the learning management system (M = 4.67)
- Findings indicate high importance in course design and communication, suggesting a need for more faculty support.
- **Statistical Significance:**
  - Weak positive relationship between perceived Importance and Ability, r(223) = .249, p < .001.
*This section highlights the strong perceptions of importance in teaching competencies and their correlation with perceived ability.*

---

## Research Question 2: Demographic Factors
- **Question:** What demographic factors are related to faculty's perceptions of the importance of course design, communication, time management, and technical competency, and their ability to teach online?
- Demographic data included: gender, academic rank, online teaching methods, and level taught.
- Descriptive statistics presented in Table 5 on page 32, Chapter IV.
- **ANOVA Results:**
  - Significant differences found in gender and level taught regarding importance of online teaching competencies.
  - No significant differences in gender, academic rank, and online teaching methods.
  - Significant differences in level taught based on perceptions of ability to teach online.
- No significant differences found in other demographic factors.
*This section provides insights into how demographic factors influence faculty perceptions of online teaching competencies.*

![diff](diff.png)
